rules/rules.js

```diff
@@ -280,54 +279,43 @@ const JSRule = function (ruleConfig) {
   // Register rule here
   if (triggers && triggers.length > 0) {
     rule.setTriggers(triggers);
-    rule = registerRule(rule);
-  }
-
-  return rule;
-};
-
-let currentProvider = automationManager;
-
-const withNewRuleProvider = function (fn) {
-  const cachedRules = [];
-  currentProvider = {
-    addRule: r => {
-      // r = factory.processRule(r);
-      // r.setConfigurationDescriptions(null);
-      // r.setConfiguration(null);
-      cachedRules.push(factory.processRule(r));
-    }
-  };
-
-  try {
-    fn();
-    const provider = factory.newRuleProvider(cachedRules);
-    osgi.registerService(provider, Java.type('org.openhab.core.automation.RuleProvider').class.getName());
-  } finally {
-    currentProvider = automationManager;
+    rule = automationManager.addRule(rule);
   }
-};
 
-const withManagedProvider = function (fn) { // eslint-disable-line no-unused-vars
-  const previousProvider = currentProvider;
-  currentProvider = automationManager;
+  // Add config to the action so that MainUI can show the script
+  const actionConfiguration = rule.actions.get(0).configuration;
+  actionConfiguration.put('type', 'application/javascript;version=ECMAScript-2021');
+  actionConfiguration.put('script', (ruleConfig.scriptCode) ? ruleConfig.scriptCode : generateUIRuleCode(ruleConfig));
 
-  try {
-    fn();
-  } finally {
-    currentProvider = previousProvider;
-  }
+  return rule;
 };
 
-const registerRule = function (rule) {
-  return currentProvider.addRule(rule);
+/**
+ * Regenerates the rule's source code to show in MainUI by parsing the rule configuration.
+ *
+ * @private
+ * @param {RuleConfig} ruleConfig The rule config describing the rule
+ * @returns {string} the regenerated source code of the rule
+ */
+const generateUIRuleCode = function (ruleConfig) {
```

Comment:
I am right in saying that this function is attempting to wrap and rewrite script code?

Surely there must be another way to do this. This is typically a very risky thing to do. For example, just at a quick glance, I suspect that this implementation suffers from a lack of proper escaping, so that scripts may contain characters which end the "execute" property of the rule object early. Fixing this escaping may be very difficult.

There will be other problems too - such as mismatching line numbers when failures occur. I would really advise against modifying raw script code, especially in JS. Much better to handle this by providing the needed metadata to the Java layer.

Reply:
> I am right in saying that this function is attempting to wrap and rewrite script code?

No, not really. This function is attempting to regenarate the actual source code of the rule to show it in the UI. But the regenerated code is never executed.

> Surely there must be another way to do this.

I hope so, because I am also not really happy with this implementation. It's just a quick proposal, if it is working that way.

I'm not sure if you've checked out the linked issue, but to summarize: For DSL rules as well as file-based Ruby rules, the roule's source code is shown in the UI, which can be really helpful when you want to know what the rule does.

To solve this in a good way, we'd basically need to access the real source code from the `.js` file, but I've no idea how to accomplish this.

Reply:
@jpg0 
I‘ve seen that Rules DSL only displays the executed code in MainUI (which makes quite sence, as this is enough to know what the rule is doing). Implementing this for JS is much easier, as we would only need to stringify the callback code and display it as the script in the UI.
Then we add a comment line that says only the execute code is displayer.

WDYT about that?